src/App.tsx

```diff
@@ -15,47 +17,130 @@ export const goodsFromServer = [
+  const visibleGoods = [...goods];
+
+  visibleGoods.sort((currentGood: string, nextGood: string) => {
+    switch (sortType) {
+      case SortType.ALPHABET:
+        return currentGood.localeCompare(nextGood);
+
+      case SortType.LENGTH:
+        return currentGood.length - nextGood.length;
+
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+}
+
+export const App: React.FC<{}> = () => {
+  const [isReversed, setReverse] = useState(false);
+  const [sortType, setSortType] = useState<SortType>(SortType.NONE);
+  const visibilityResetButton = sortType !== SortType.NONE || isReversed;
+  const visibleItems = getReorderedGoods(
+    goodsFromServer,
+    sortType,
+    isReversed,
+  );
+
+  const sortByAlphabet = () => {
+    setSortType(SortType.ALPHABET);
+  };
+
+  const sortByLength = () => {
+    setSortType(SortType.LENGTH);
+  };
+
+  const reverseList = () => {
+    setReverse(!isReversed);
+  };
+
+  const resetList = () => {
+    setSortType(SortType.NONE);
+    setReverse(false);
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
           type="button"
-          className="button is-info is-light"
+          onClick={sortByAlphabet}
+          className={classNames(
+            'button',
+            'is-info',
+            { 'is-light': sortType !== SortType.ALPHABET },
+          )}
         >
           Sort alphabetically
         </button>
 
         <button
           type="button"
-          className="button is-success is-light"
+          onClick={sortByLength}
+          className={classNames(
+            'button',
+            'is-success',
+            { 'is-light': sortType !== SortType.LENGTH },
+          )}
         >
           Sort by length
         </button>
 
         <button
           type="button"
-          className="button is-warning is-light"
+          onClick={() => {
+            reverseList();
+          }}
+          className={classNames(
+            'button',
+            'is-warning',
+            {
+              'is-light': !isReversed,
+            },
+          )}
         >
           Reverse
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {visibilityResetButton && (
+          <button
+            type="button"
+            onClick={resetList}
+            className="button is-danger is-light"
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
         <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
+          {visibleItems.map(good => (
+            <li
+              key={uuidv4()}
```

Comment:
the key should be unique but not random. Now we rerender the list each time because each time we have new key there